crates/lox_core/src/time/leap_seconds/api.rs

```diff
@@ -0,0 +1,156 @@
+use crate::time::leap_seconds::gen::{LEAP_SECONDS, LS_EPOCHS};
+
+const MJD_EPOCH: f64 = 2400000.5;
+const SECONDS_PER_DAY: f64 = 86400.0;
+
+// Constants for calculating the offset between TAI and UTC for
+// dates between 1960-01-01 and 1972-01-01
+// See ftp://maia.usno.navy.mil/ser7/tai-utc.dat
+
+const EPOCHS: [u64; 14] = [
+    36934, 37300, 37512, 37665, 38334, 38395, 38486, 38639, 38761, 38820, 38942, 39004, 39126,
+    39887,
+];
+
+const OFFSETS: [f64; 14] = [
+    1.417818, 1.422818, 1.372818, 1.845858, 1.945858, 3.240130, 3.340130, 3.440130, 3.540130,
+    3.640130, 3.740130, 3.840130, 4.313170, 4.213170,
+];
+
+const DRIFT_EPOCHS: [u64; 14] = [
+    37300, 37300, 37300, 37665, 37665, 38761, 38761, 38761, 38761, 38761, 38761, 38761, 39126,
+    39126,
+];
+
+const DRIFT_RATES: [f64; 14] = [
+    0.0012960, 0.0012960, 0.0012960, 0.0011232, 0.0011232, 0.0012960, 0.0012960, 0.0012960,
+    0.0012960, 0.0012960, 0.0012960, 0.0012960, 0.0025920, 0.0025920,
+];
+
+#[derive(PartialEq, Debug)]
+pub enum LeapSecondError {
+    // UTC is not defined for dates before 1960-01-01.
+    UTCDateBefore1960,
+    UTCDateOutOfRange,
+}
+
+type TwoPartDateTime = (f64, f64);
+
+fn is_sorted(array: &[u64]) -> bool {
+    array.windows(2).all(|x| x[0] <= x[1])
+}
+
+fn leap_seconds(mjd: f64) -> Result<f64, LeapSecondError> {
+    // Invariant: LS_EPOCHS must be sorted for the search below to work
+    assert!(is_sorted(&LS_EPOCHS));
+
+    let threshold = mjd.floor() as u64;
+    let position = LS_EPOCHS
+        .iter()
+        .rposition(|item| item <= &threshold)
+        .ok_or(LeapSecondError::UTCDateOutOfRange)?;
+
+    Ok(LEAP_SECONDS[position])
+}
+
+/// Returns the difference between UTC and TAI for a given date
+///
+/// Input is a two-part UTC Julian datetime for increased precision. The first part is
+/// the UTC day number, and the second part is the second offset
+pub fn offset_utc_tai(utc_date_time: TwoPartDateTime) -> Result<f64, LeapSecondError> {
+    // This function uses the [ERFA convention](https://github.com/liberfa/erfa/blob/master/src/dtf2d.c#L49)
+    // for Julian day numbers representing UTC dates during leap seconds.
+    let mjd = utc_date_time.0 - MJD_EPOCH + utc_date_time.1;
+
+    // Before 1960-01-01
+    if mjd < 36934.0 {
```

Comment:
Practically yes, but only because they have the same hidden dependency - UTC starts in 1 jan 1960 so there's no way of defining `EPOCHS` before that. That being said, I guess a meaningful EPOCHS will always start at 1 jan 1960.

Any inputs @helgee ?

Reply:
I realize that this is rather late (apologies @matzipan) but I am strongly considering dropping the 1960-1972 part of the code. These USNO UTC numbers are not generally accepted (see [here](https://twitter.com/helge_e/status/1236780104984002565)) and we do not really care about historical accuracy for our use cases anyway (unlike SOFA).
If we do this, we could also model leap seconds as integers instead of `f64` which would be a plus IMHO.

Reply:
> These USNO UTC numbers are not generally accepted (see [here](https://twitter.com/helge_e/status/1236780104984002565))

I don't have twitter anymore so can't read the thread :( 

>  I am strongly considering dropping the 1960-1972 part of the code.

So if I understand correctly you are suggesting to drop the error for `UTCDateBefore1960`? What would be the 1972 part of the change?

Reply:
> These USNO UTC numbers are not generally accepted (see [here](https://twitter.com/helge_e/status/1236780104984002565))

I don't have twitter anymore so can't read the thread :( 

>  I am strongly considering dropping the 1960-1972 part of the code.

So if I understand correctly you are suggesting to drop the error for `UTCDateBefore1960`? What would be the 1972 part of the change?

Reply:
https://github.com/lox-space/lox/issues/32